x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/FrozenIndexInput.java

```diff
@@ -591,33 +324,32 @@ private void writeCacheFile(
         while (remaining > 0L) {
             final int bytesRead = readSafe(input, copyBuffer, relativePos, end, remaining, frozenCacheFile);
             if (bytesRead > buf.remaining()) {
+                // always fill up buf to the max
+                final int bytesToAdd = buf.remaining();
+                buf.put(copyBuffer, 0, bytesToAdd);
+                assert buf.remaining() == 0;
                 long bytesWritten = positionalWrite(fc, fileChannelPos + bytesCopied, buf);
                 bytesCopied += bytesWritten;
                 progressUpdater.accept(bytesCopied);
+                // add the remaining bytes to buf
+                buf.put(copyBuffer, bytesToAdd, bytesRead - bytesToAdd);
+            } else {
+                buf.put(copyBuffer, 0, bytesRead);
             }
-            buf.put(copyBuffer, 0, bytesRead);
             remaining -= bytesRead;
         }
+        // ensure that last write is aligned on 4k boundaries (= page size)
+        final int off = buf.position() % SharedBytes.PAGE_SIZE;
+        final int adjustment = off == 0 ? 0 : SharedBytes.PAGE_SIZE - off;
+        buf.position(buf.position() + adjustment);
         long bytesWritten = positionalWrite(fc, fileChannelPos + bytesCopied, buf);
-        bytesCopied += bytesWritten;
+        bytesCopied += (bytesWritten - adjustment); // adjust to not break RangeFileTracker
```

Comment:
Maybe the stats should show the exact bytes copied (ie we can confirm the 4K alignment) but progress updater got updated with the adjusted bytes copied?

Reply:
++

Reply:
I will defer that to a follow-up as it will break a lot of tests (again)